### ISSUE TITLE
Few questions in regards to the experimental extension system

### ISSUE THREAD

**Issue by qballer** [teambit/bit#1919]
### Description

I'm working on a new version of the envs and I have a few questions/requests: 

1. How do you pass command line arguments to a compiler ? When we want to change the choice of the configuration/overrides we should be able to supply a way for the compiler to accept and expose CLI arguments. 

2. How does bit overrides play with `getDynamicDependencies` ?  I'm assuming some kind of merge between them would make sense where override is more significant.

3. Can you supply `d.ts` files for bit's API ? When extending bit with the extension system or using the capsule API, it would great to have type informatio on those functions. related to #1753 

5. Create workspace for testing. On some e2e testing I would like to create a workspace for testing that the compiled component is working. I know bit holds those capabilities, I wonder if you can offer them as a test API?

**Comment by davidfirst** [comment]
1. I don't think this option is available, you can change your workspace config of the compiler to have all options you need. Can you provide an example when an option is needed from the CLI and can't be done by changing the config?
2. Yes. In case a component has its own configuration of `getDynamicDependencies` and the workspace has a different configuration for that component, the objects will be merged by JS `Object.assign()` method. The more specific always wins.
3. OK, we can work on it, I don't see why not.
4. Not sure what you mean by test API. It can be easily done now (and I do it all the time). Just mark the test you want as `.only` and run the tests as following: `npm run e2e-test --debug --keep_envs`. These flags ensure two critical things: 1) it will not delete the workspace once the test is completed. 2) it prints the path to the workspace so you could easily cd into it and continue working.

**Comment by qballer** [comment]
1. It's when you don't want to change the config and just have it run like that for this run. 
2. Overrides are merged over the compiler one, right ? 
3. This would be very cool! I have to debug and see. 
4. I want to write an automated e2e test where I create a directory with a bit component and bit build with the compiler I'm testing. This is kind of a capsule with bit init from an object. I hope I explained it better.

**Comment by davidfirst** [comment]
1. I guess this will be part of the Extensions feature. As a compiler extension, you'll be able to hook into `bit build` command and add flags.
2. Erase what I said before. The `Object.assign()` is done for every property in `overrides` except `env`. For environment settings (compiler/tester) we don't look into the specific objects. If a component has its settings of a compiler, Bit will use it and will not try to merge settings from the workspace. Only when `env.compiler` is not set on the component level, it'll use the workspace.
The strategies of how the compiler/tester is loaded are described in `env-extension.js` file as follows:

> * load the compiler/tester according to the following strategies:
>    * 1. from component config (bit.json/package.json of the component) if it was written.
>    * 2. from component model. an imported component might not have the config written.
>    * for author, it's irrelevant, because upon import it's written to consumer config (if changed).
>    * 3. from consumer config overrides. (bit.json/package.json of the consumer when this component
>    * overrides the general env config).
>    * 4. from consumer config.
4. I think I got it, I see how it could be useful, but you can achieve it easily now.

**Comment by qballer** [comment]
2. Will look more in to it and provide feedback soon.
3. How long for this ? 
4. Can you provide a code example ?

**Comment by davidfirst** [comment]
3. let's take it offline.
4. look into `e2e` directory, you'll see tons of examples. There is no dedicated Bit command for this. They just create a directory in your temp dir, run `bit init`, add files as needed and run the rest of Bit commands in that directory. Once done, the directory is deleted. If needed to work against a remote scope, they create another directory and run `bit init --bare` on that dir. Let's discuss if that's not clear or you need something else.

**Comment by GiladShoham** [comment]
I first suggest we will separate each one of those to a different issue.
It's becoming very hard to follow on what going on here since that's a bunch of unrelated issues and each time someone needs to go up and down to see what each number represents.
for specific comments about the issues.

1. For now, we just call the compiler action directly, so as compiler writer you have access to the process.env. If you want to pass it on to a child process / capsule what ever you can do it as of now.
For tester's, it's a bit different since they are running on a child process. however, we just push all the envs vars into the child process. so as a tester writer you have access to the process.env as well (same as a compiler for you).

2. @davidfirst I'm not sure you were followed about what @qballer asked here. with the new envs API we provide a way for an env to add additional deps to the component. that to support a case for example:
you have a babel compiler which gets config about plugins, the compiler will parse those plugins and will add the corresponding packages to the component packages deps. That's will go into compilerPackageDependencies / testerPackageDependencies in the model. As I understand the question is about the package.json of the component will it include both overrides.dependencies values and the compilerPackageDependencies as well?
From what I see in the code [here](https://github.com/teambit/bit/blob/master/src/consumer/component/package-json-file.js#L84)
I believe they will be merged (assuming that the devDependencies including the overrides deps, but I'm not sure about it. @davidfirst can you confirm it? where is the code that handles the overrides when writing a component?).

3. We should estimate it. I don't know how much time exactly it will take. @qballer feel free to help us with this, as you already know the API as much as us, and better with TS:)

4. I think that what @qballer wants is something like release our e2e helper as an SDK, so as compiler / tester writer, you will be able to e2e test it without requiring 'bit-bin' / change the bit-bin itself or copy-paste half of the e2e-helper code into your env repo.
Maybe we can isolate it a bit more, and release it as a bit component / components. that's a pretty good idea and I think not too hard.

**Comment by qballer** [comment]
@GiladShoham you understand me :)
Separate issue are hard to track, and I want this to be successful in improving our envs. Hence the single issue. This is in order to improve the envs UX . if you want to break them up I won't stop you :). 
  
1. args: granted my example was in regards to environment variables, but I would also like to pass command line arguments. Any way to do that?
3. typed API: I don't mind doing the PR. This won't be simple to provide good user experience 
4. YES. this is so important here.

**Comment by davidfirst** [comment]
I think next time I'll let Gilad answering first, you guys understand each other better :)

2. As far as I remember neither are correct. It doesn't write it into `compilerPackageDependencies`, nor add it to `overrides`. The tester/compiler packages are going into `devPackages`. The overrides are only when you originally changed the `overrides` property in the workspace or directly in the `package.json` of the component.

**Comment by GiladShoham** [comment]
@qballer
1. There is no such way right now. but we can make one. I think about maybe a syntax similar to [npm scripts](https://docs.npmjs.com/cli/run-script)
something like:
bit build [my-comp] [--no-cache (bit's build args) -- --compiler-arg compiler-val 
with the -- as a separator between bit's args and the compiler args.
(We will have to see how complicated is to parse it with commander parser).
2. @davidfirst It should have been saved into compilerPackageDependencies in the models, and into the devDeps in the package.json. what we are wondering is what the devDeps will include in case of copmilerPackageDeps and override devDeps.

**Comment by davidfirst** [comment]
2. still not sure I follow, but if you did change the `overrides` property and mark a devDependency with `-`, it will be removed from the component package.json. The `overrides` doesn't have compilerPackagesDependencies, instead, it uses devDependencies for that (same as package.json).

**Comment by qballer** [comment]
@qballer 
in most shells -- already inherits all arguments.

**Comment by davidfirst** [comment]
4. Done. I exposed the e2e-helper class to the API. You can use it as follows:
```
const api = require('./dist/api');
const helper = new api.e2eTestHelper();
helper.setNewLocalAndRemoteScopes();
helper.createFile('bar', 'foo.js', '');
helper.tagAllComponents();
helper.exportAllComponents();
helper.reInitLocalScope();
helper.addRemoteScope();
helper.importComponent('bar/foo');
```

**Comment by GiladShoham** [comment]
@davidfirst @qballer 
Just a quick fix.
you can use 
const api = require('bit-bin');
personally, I would call it:
const bit = require('bit-bin');

**Comment by qballer** [comment]
That's awesome, I will need a build and a test function on that helper.

**Comment by GiladShoham** [comment]
@qballer The refactor for the e2e-helper is merged into master.
